### PR TITLE
Mark flow elephant/v2

### PR DESCRIPTION
**Pull request by inashivb**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/5647

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2130

Previous PR: https://github.com/OISF/suricata/pull/11645

Changes since v1:
- added a rule keyword besides allowing flow rate config in suricata.yaml

### CODE REVIEW

File: src/detect-flow-rate.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    uint64_t age = SCTIME_SECS(p->flow->lastts) - SCTIME_SECS(p->flow->startts);
+    uint64_t rate = (p->flow->tosrcbytecnt + p->flow->todstbytecnt) / age;
+
+    uint64_t expected_rate = (uint64_t)ctx;
+    return rate == expected_rate;
+}
+
+static int DetectFlowRateSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+{
+    uint64_t rate = atoll(rawstr);
```
Comment (victorjulien) [inline src/detect-flow-rate.c:41]: will need to use the `StringParseUint64` or `StringParseU64RangeCheck`

**Comment by inashivb** [comment]
Need to find a way to keep time blocked data structures for flow management so this can be done on any interval. Victor suggested we could start w this trivial keyword though. Interested in hearing if it makes sense to others too.

File: src/flow.h [hunk 2]
```diff
@@ -499,6 +501,7 @@ typedef struct Flow_
     uint64_t todstbytecnt;
     uint64_t tosrcbytecnt;
 
+    bool elephant;
```
Comment (victorjulien) [inline src/flow.h:504]: how does this affect the flow memory layout? (pahole)

File: src/detect-flow-rate.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        return 0;
+    }
+    uint64_t age = SCTIME_SECS(p->flow->lastts) - SCTIME_SECS(p->flow->startts);
+    uint64_t rate = (p->flow->tosrcbytecnt + p->flow->todstbytecnt) / age;
+
+    uint64_t expected_rate = (uint64_t)ctx;
+    return rate == expected_rate;
+}
+
+static int DetectFlowRateSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
```
Comment (victorjulien) [inline src/detect-flow-rate.c:39]: I think we need to think about how the arguments will express time later.

File: src/detect-flow-rate.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+#include "suricata-common.h"
+#include "rust.h"
+#include "detect-flow-rate.h"
+#include "detect-engine.h"
+#include "detect-engine-prefilter.h"
+#include "detect-engine-uint.h"
+#include "detect-parse.h"
+
+static int DetectFlowRateMatch(
+        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
```
Comment (victorjulien) [inline src/detect-flow-rate.c:27]: I would expect this to use the detect int handling code

**Review by victorjulien — CHANGES_REQUESTED** [review]
I would like to see this split between a PR for the flow.rate and the elephant marking.

Also please add user docs.

Other comments inline.

File: src/detect-flow-rate.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+{
+    if (p->flow == NULL) {
+        return 0;
+    }
+    uint64_t age = SCTIME_SECS(p->flow->lastts) - SCTIME_SECS(p->flow->startts);
+    uint64_t rate = (p->flow->tosrcbytecnt + p->flow->todstbytecnt) / age;
+
+    uint64_t expected_rate = (uint64_t)ctx;
+    return rate == expected_rate;
```
Comment (inashivb) [inline src/detect-flow-rate.c:36]: incorrect. should'nt look for exact match but greater than equal to instead

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.flow.dcerpc_tcp | 40 | 42 | 105.0% | 

Pipeline 23324

File: src/detect-flow-rate.c (hunk 3 above)
Comment (inashivb) [reply src/detect-flow-rate.c:39]: Making a struct to easily extend later on

File: src/flow.h (hunk 2 above)
Comment (inashivb) [reply src/flow.h:504]: This was creating a 7 byte hole. Moved it above where there was already a 2 byte hole which is now reduced to 1 byte

**Pull request closed without merging**